tools/optimizer-server/src/main.rs

```diff
@@ -163,36 +164,41 @@ fn generate_event_info(path: &Path) -> Result<EventInfo, io::Error> {
     })
 }
 
-fn send_event_info(event_info: &Vec<EventInfo>) -> Result<(), SendError> {
+fn send_event(event: &EventInfo) -> Result<(), SendError> {
     let mut writer = io::stdout();
-    serde_json::to_writer(writer.lock(), event_info).map_err(SendError::Serde)?;
+    let event_string = serde_json::to_string(event).map_err(SendError::Serde)?;
+    writer
+        .write_all(format!("{event_string}\n").as_bytes())
+        .map_err(SendError::IO)?;
     writer.flush().map_err(SendError::IO)
 }
 
+fn handle_event(event: &FanotifyEvent, event_duplicate: &mut Vec<String>) -> Result<(), SendError> {
+    let path = get_fd_path(event.fd).map_err(SendError::IO)?;
+    let info = generate_event_info(&path).map_err(SendError::IO)?;
+    if !event_duplicate.contains(&info.path) {
+        send_event(&info)?;
+        event_duplicate.push(info.path);
+    }
+    Ok(())
+}
+
 fn handle_fanotify_event(fd: i32) {
     let mut fds = [PollFd::new(fd.as_raw_fd(), PollFlags::POLLIN)];
-    let event_info = Arc::new(Mutex::<Vec<EventInfo>>::new(Vec::default()));
-    let local_event_info = event_info.clone();
-
-    if let Err(e) = Signals::new([SIGTERM]).map(|mut signals| {
-        thread::spawn(move || {
-            for _ in signals.forever() {
-                eprintln!("received SIGTERM signal, sending event information");
-                if let Ok(event_info) = local_event_info.lock().as_deref() {
-                    if let Err(e) = send_event_info(event_info) {
-                        eprintln!("failed to send event information {event_info:?} {e:?}");
-                    }
-                }
-                eprintln!("send event information successfully, exiting");
-                process::exit(0);
-            }
-        })
-    }) {
+    let mut event_duplicate = Vec::new();
+
+    let term = Arc::new(AtomicBool::new(false));
+    if let Err(e) = signal_hook::flag::register(signal_hook::consts::SIGTERM, Arc::clone(&term)) {
         eprintln!("failed to set SIGTERM signal handler {e:?}");
         return;
     }
 
     loop {
+        if term.load(Ordering::Relaxed) {
```

Comment:
`Ordering::Relaxed` is not reliable in the case breaking this loop

Reply:
Should also wake up the Poller when receiving a TERM signal

Reply:
Fixed.